indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java

```diff
@@ -287,90 +292,347 @@ public TaskLock lock(final Task task, final Interval interval, long timeoutMs) t
           // timekeeping other than the wall clock.
           version = DateTimes.nowUtc().toString();
         }
 
-        posseToUse = new TaskLockPosse(new TaskLock(task.getGroupId(), dataSource, interval, version));
-        running.get(dataSource)
-               .put(interval, posseToUse);
+      final TaskLockPosse posseToUse = new TaskLockPosse(
+          new TaskLock(lockType, groupId, dataSource, interval, version, priority)
+      );
+      running.computeIfAbsent(dataSource, k -> new TreeMap<>(Comparators.intervalsByStartThenEnd()))
+             .computeIfAbsent(interval, k -> new ArrayList<>())
+             .add(posseToUse);
+
+      return posseToUse;
+    }
+    finally {
+      giant.unlock();
+    }
+  }
+
+  /**
+   * Upgrade a lock for the given task and interval if possible. Shared locks and revoked locks cannot be upgraded.
+   *
+   * @return a succeeded {@link LockResult} if the lock is upgraded. If an upgrade for a revoked lock is requested, a
+   * failed {@link LockResult} is returned.
+   *
+   * @throws IllegalStateException if an upgrade for a shared lock is requested
+   */
+  public LockResult upgrade(Task task, Interval interval)
+  {
+    giant.lock();
+
+    try {
+      final TaskLockPosse taskLockPosseToUpdate = getOnlyTaskLockPosse(
+          task,
+          interval,
+          findLockPossesContainingInterval(task.getDataSource(), interval)
+      );
+
+      final TaskLock lock = taskLockPosseToUpdate.getTaskLock();
+
+      Preconditions.checkState(
+          lock.getType().equals(TaskLockType.EXCLUSIVE),
+          "Shared lock cannot be upgraded"
+      );
+
+      if (lock.isRevoked()) {
+        return LockResult.fail(true);
+      } else {
+        log.info("Upgrading taskLock[%s] for task[%s] and interval[%s]", lock, task.getId(), interval);
+
+        if (lock.isUpgraded()) {
+          log.warn("TaskLock[%s] is already upgraded", lock);
+          return LockResult.ok(lock);
+        } else {
+          final TaskLock upgradedLock = lock.upgrade();
+          taskStorage.replaceLock(task.getId(), lock, upgradedLock);
+
+          // Possible exceptions are already checked above
+          final List<TaskLockPosse> possesHolder = running.get(task.getDataSource()).get(lock.getInterval());
 
-        log.info("Created new TaskLockPosse: %s", posseToUse);
+          possesHolder.remove(taskLockPosseToUpdate);
+          possesHolder.add(taskLockPosseToUpdate.withTaskLock(upgradedLock));
+          log.info("Upgraded taskLock[%s]");
+          return LockResult.ok(upgradedLock);
+        }
       }
+    }
+    finally {
+      giant.unlock();
+    }
+  }
 
-      return posseToUse;
+  /**
+   * Downgrade a lock for the given task and interval if possible. Shared locks cannot be downgraded.
+   *
+   * @return a downgraded lock if succeeded.
+   *
+   * @throws IllegalStateException if a downgrade for a shared lock is requested
+   */
+  public TaskLock downgrade(Task task, Interval interval)
```

Comment:
Since multiple tasks can obtain the same lock, is this vulnerable to a problem like:

- Task A and B share a lock
- Task A enters a critical section and upgrades the lock
- Task B enters a critical section and upgrades the lock
- Task A downgrades the lock
- Now Task B's lock is downgraded even though it's still in a critical section

Reply:
Good point. I guess Task A and B are subtasks of a task because it's allowed to share an exclusive lock only if they have the same groupId. In this case, upgrade/downgrade must be atomic.

BTW, while I'm looking into this, I realized that the atomic operation in a critical section can be achieved without complicated upgrade and downgrade. So, in the latest patch, I removed them and added a new method ```doInCriticalSection()``` to ```TaskLockBox``` instead.

```doInCriticalSection()```method performs a given action with a guarantee that the locks of the task requested the action are not preempted while performing the action. Currently, this is guaranteed by acquiring a ReentrantLock, which means every call to methods of TaskLockBox are blocked until doInCriticalSection() is finished. I think this should be fine because 
- Publishing segments should not take a long time.
- Publishing segments won't be frequent even after the fine-grained locking proposed in https://github.com/druid-io/druid/issues/4479 is implemented. The compaction task will be the one which needs more frequent locking and publishing segments than other task types, but it should be suppressed to not make a big overhead to the cluster.